File: rdkit-sys/src/bridge/scaffold_network.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    // Trick from https://github.com/dtolnay/cxx/issues/741
    // pub struct ROMolSharedPtr {
    //     ptr: SharedPtr<ROMol>,
    // }

    unsafe extern "C++" {
        include!("wrapper/include/scaffold_network.h");

        pub type ScaffoldNetworkParamsLocal;
        pub type ROMol = crate::ro_mol_ffi::ROMol;

        pub fn default_scaffold_network_params() -> SharedPtr<ScaffoldNetworkParamsLocal>;

        pub fn new_scaffold_network_params(
            bond_breaker_smarts: &Vec<String>,
        ) -> SharedPtr<ScaffoldNetworkParamsLocal>;

        pub fn set_include_generic_scaffolds(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_include_generic_bond_scaffolds(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_include_scaffolds_without_attachments(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_include_scaffolds_with_attachments(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_keep_only_first_fragment(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_prune_before_fragmenting(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_flatten_isotopes(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_flatten_chirality(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_flatten_keep_largest(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );
        pub fn set_collect_mol_counts(
            params: &mut SharedPtr<ScaffoldNetworkParamsLocal>,
            input: bool,
        );

        pub fn get_include_generic_scaffolds(
            params: &SharedPtr<ScaffoldNetworkParamsLocal>,
        ) -> bool;
        pub fn get_include_generic_bond_scaffolds(
            params: &SharedPtr<ScaffoldNetworkParamsLocal>,
        ) -> bool;
        pub fn get_include_scaffolds_without_attachments(
            params: &SharedPtr<ScaffoldNetworkParamsLocal>,
        ) -> bool;
        pub fn get_include_scaffolds_with_attachments(
            params: &SharedPtr<ScaffoldNetworkParamsLocal>,
        ) -> bool;
        pub fn get_keep_only_first_fragment(params: &SharedPtr<ScaffoldNetworkParamsLocal>)
            -> bool;
        pub fn get_prune_before_fragmenting(params: &SharedPtr<ScaffoldNetworkParamsLocal>)
            -> bool;
        pub fn get_flatten_isotopes(params: &SharedPtr<ScaffoldNetworkParamsLocal>) -> bool;
        pub fn get_flatten_chirality(params: &SharedPtr<ScaffoldNetworkParamsLocal>) -> bool;
        pub fn get_flatten_keep_largest(params: &SharedPtr<ScaffoldNetworkParamsLocal>) -> bool;
        pub fn get_collect_mol_counts(params: &SharedPtr<ScaffoldNetworkParamsLocal>) -> bool;

        pub type ScaffoldNetworkClass;
        pub fn default_scaffold_network() -> SharedPtr<ScaffoldNetworkClass>;

        // pub fn create_scaffold_network(
        //     mol: &CxxVector<ROMolSharedPtr>,
        //     scaffold_network_params: &SharedPtr<ScaffoldNetworkParamsLocal>,
        // ) -> SharedPtr<ScaffoldNetworkClass>;

        // pub fn update_scaffold_network(
        //     scaffold_network: &mut SharedPtr<ScaffoldNetworkClass>,
        //     mol: &CxxVector<ROMolSharedPtr>,
        //     scaffold_network_params: &SharedPtr<ScaffoldNetworkParamsLocal>,
        // );
    }
}

File: rdkit-sys/wrapper/src/scaffold_network.cc
#include "rust/cxx.h"
#include <GraphMol/GraphMol.h>
#include <GraphMol/ScaffoldNetwork/ScaffoldNetwork.h>

namespace RDKit {
  using ScaffoldNetworkParams = ScaffoldNetwork::ScaffoldNetworkParams;

  std::shared_ptr<ScaffoldNetworkParams> default_scaffold_network_params() {
    ScaffoldNetworkParams *network_params = new ScaffoldNetworkParams();
    return std::shared_ptr<ScaffoldNetworkParams>(network_params);
  }

  std::shared_ptr<ScaffoldNetworkParams> new_scaffold_network_params(const rust::Vec<rust::String> &bondBreakersSmarts) {
    std::vector<std::string> cc_vec;

    for (auto smarts : bondBreakersSmarts) {
      cc_vec.push_back(std::string(smarts));
    }

    ScaffoldNetworkParams *network_params = new ScaffoldNetworkParams(cc_vec);
    return std::shared_ptr<ScaffoldNetworkParams>(network_params);
  }

  void set_include_generic_scaffolds(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->includeGenericScaffolds = input;
  }

  void set_include_generic_bond_scaffolds(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->includeGenericBondScaffolds = input;
  }

  void set_include_scaffolds_without_attachments(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->includeScaffoldsWithoutAttachments = input;
  }

  void set_include_scaffolds_with_attachments(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->includeScaffoldsWithAttachments = input;
  }

  void set_keep_only_first_fragment(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->keepOnlyFirstFragment = input;
  }

  void set_prune_before_fragmenting(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->pruneBeforeFragmenting = input;
  }

  void set_flatten_isotopes(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->flattenIsotopes = input;
  }

  void set_flatten_chirality(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->flattenChirality = input;
  }

  void set_flatten_keep_largest(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->flattenKeepLargest = input;
  }

  void set_collect_mol_counts(std::shared_ptr<ScaffoldNetworkParams> &params, bool input) {
    params->collectMolCounts = input;
  }

  using ScaffoldNetworkClass = ScaffoldNetwork::ScaffoldNetwork;

  std::shared_ptr<ScaffoldNetworkClass> default_scaffold_network() {
    ScaffoldNetworkClass *scaffold_network = new ScaffoldNetworkClass();
    return std::shared_ptr<ScaffoldNetworkClass>(scaffold_network);
  }

  bool get_include_generic_scaffolds(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->includeGenericScaffolds;
  }
  bool get_include_generic_bond_scaffolds(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->includeGenericBondScaffolds;
  }
  bool get_include_scaffolds_without_attachments(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->includeScaffoldsWithoutAttachments;
  }
  bool get_include_scaffolds_with_attachments(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->includeScaffoldsWithAttachments;
  }
  bool get_keep_only_first_fragment(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->keepOnlyFirstFragment;
  }
  bool get_prune_before_fragmenting(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->pruneBeforeFragmenting;
  }
  bool get_flatten_isotopes(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->flattenIsotopes;
  }
  bool get_flatten_chirality(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->flattenChirality;
  }
  bool get_flatten_keep_largest(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->flattenKeepLargest;
  }
  bool get_collect_mol_counts(const std::shared_ptr<ScaffoldNetworkParams> &params) {
    return params->collectMolCounts;
  }

  void update_scaffold_network(const std::shared_ptr<ROMol> &mol, std::shared_ptr<ScaffoldNetworkClass> &scaffold_network, const std::shared_ptr<ScaffoldNetworkParams> &scaffold_network_params) {
    std::vector<ROMol> mols;
    mols.push_back(*mol);
    ScaffoldNetwork::updateScaffoldNetwork(&mols, *scaffold_network, *scaffold_network_params);
  }

}